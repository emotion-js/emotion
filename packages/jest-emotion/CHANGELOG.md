# jest-emotion

## 10.0.32

### Patch Changes

- [`5af2755`](https://github.com/emotion-js/emotion/commit/5af27554d3aec20691a4675e57433bb01cd535a4) [#1808](https://github.com/emotion-js/emotion/pull/1808) Thanks [@Andarist](https://github.com/Andarist)! - Fixed compatibility with latest `jest` TS types.

## 10.0.27

### Patch Changes

- [`4c62ae9`](https://github.com/emotion-js/emotion/commit/4c62ae9447959d438928e1a26f76f1487983c968) [#1698](https://github.com/emotion-js/emotion/pull/1698) Thanks [@Andarist](https://github.com/Andarist)! - Add LICENSE file
- Updated dependencies [[`4c62ae9`](https://github.com/emotion-js/emotion/commit/4c62ae9447959d438928e1a26f76f1487983c968)]:
  - @emotion/core@10.0.27
  - emotion@10.0.27

## 10.0.26

### Patch Changes

- [`b3c5b8de`](https://github.com/emotion-js/emotion/commit/b3c5b8de66e42dd2e6459862c9603f012ba01d54) [#1667](https://github.com/emotion-js/emotion/pull/1667) Thanks [@Andarist](https://github.com/Andarist)! - Revert improved support for Enzyme's shallow rendering - its release was an unforseen breaking change.

## 10.0.25

### Patch Changes

- [`858c6e70`](https://github.com/emotion-js/emotion/commit/858c6e70e2aa83d159dba00af16f1e34a6d93fd0) [#1648](https://github.com/emotion-js/emotion/pull/1648) Thanks [@ajs139](https://github.com/ajs139)! - Improve support for Enzyme's shallow rendering.

## 10.0.17

### Patch Changes

- [fa5ffa80](https://github.com/emotion-js/emotion/commit/fa5ffa80890a079a3f333a463284bdb683cc2e0c) [#1477](https://github.com/emotion-js/emotion/pull/1477) Thanks [@liamcmitchell-sc](https://github.com/liamcmitchell-sc)! - Fixed snapshots when using Enzyme serializer and its deep mode in combination with fragments
  - [c8de890b](https://github.com/emotion-js/emotion/commit/c8de890b5aeeaafdcc5742aab310a61aebd666bf) [#1486](https://github.com/emotion-js/emotion/pull/1486) Thanks [@Andarist](https://github.com/Andarist)! - Fix printing speedy (usually this means production) styles
  - [66cda641](https://github.com/emotion-js/emotion/commit/66cda64128631790b81e3c9df273a972358ea593) [#1478](https://github.com/emotion-js/emotion/pull/1478) Thanks [@Andarist](https://github.com/Andarist)! - Update Babel dependencies

## 10.0.14

### Patch Changes

- [c0eb604d](https://github.com/emotion-js/emotion/commit/c0eb604d) [#1419](https://github.com/emotion-js/emotion/pull/1419) Thanks [@mitchellhamilton](https://github.com/mitchellhamilton)! - Update build tool

## 10.0.11

### Patch Changes

- [3705706](https://github.com/emotion-js/emotion/commit/3705706) [#1344](https://github.com/emotion-js/emotion/pull/1292) Thanks [@alanctkc](https://github.com/alanctkc)! - Allow jest-emotion matcher to match target by RegExp
