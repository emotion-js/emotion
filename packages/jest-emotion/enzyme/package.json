{
  "main": "dist/jest-emotion.cjs.js",
  "module": "dist/jest-emotion.esm.js",
  "types": "../types/enzyme",
  "preconstruct": {
    "source": "../src/enzyme"
  }
}
