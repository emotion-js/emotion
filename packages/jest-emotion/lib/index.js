'use strict';

exports.__esModule = true;
exports.getStyles = getStyles;
exports.createSerializer = createSerializer;

var _css = require('css');

var css = _interopRequireWildcard(_css);

var _replaceClassNames = require('./replace-class-names');

function _interopRequireWildcard(obj) { if (obj && obj.__esModule) { return obj; } else { var newObj = {}; if (obj != null) { for (var key in obj) { if (Object.prototype.hasOwnProperty.call(obj, key)) newObj[key] = obj[key]; } } newObj.default = obj; return newObj; } }

function getNodes(node) {
  var nodes = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : [];

  if (node.children) {
    for (var _iterator = node.children, _isArray = Array.isArray(_iterator), _i = 0, _iterator = _isArray ? _iterator : _iterator[Symbol.iterator]();;) {
      var _ref;

      if (_isArray) {
        if (_i >= _iterator.length) break;
        _ref = _iterator[_i++];
      } else {
        _i = _iterator.next();
        if (_i.done) break;
        _ref = _i.value;
      }

      var child = _ref;

      getNodes(child, nodes);
    }
  }

  if (typeof node === 'object') {
    nodes.push(node);
  }

  return nodes;
}

function getClassNames(selectors, classes) {
  return classes ? selectors.concat(classes.split(' ')) : selectors;
}

function getClassNamesFromProps(selectors, props) {
  return getClassNames(selectors, props.className || props.class);
}

function getClassNamesFromDOMElement(selectors, node) {
  return getClassNames(selectors, node.getAttribute('class'));
}

function getClassNamesFromNodes(nodes) {
  return nodes.reduce(function (selectors, node) {
    return isReactElement(node) ? getClassNamesFromProps(selectors, node.props) : getClassNamesFromDOMElement(selectors, node);
  }, []);
}

function getStyles(emotion) {
  return Object.keys(emotion.caches.inserted).reduce(function (style, current) {
    if (emotion.caches.inserted[current] === true) {
      return style;
    }
    return style + emotion.caches.inserted[current];
  }, '');
}

function isReactElement(val) {
  return val.$$typeof === Symbol.for('react.test.json');
}

var domElementPattern = /^((HTML|SVG)\w*)?Element$/;

function isDOMElement(val) {
  return val.nodeType === 1 && val.constructor && val.constructor.name && domElementPattern.test(val.constructor.name);
}

function createSerializer(emotion) {
  var _ref2 = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : {},
      classNameReplacer = _ref2.classNameReplacer,
      _ref2$DOMElements = _ref2.DOMElements,
      DOMElements = _ref2$DOMElements === undefined ? true : _ref2$DOMElements;

  function print(val, printer) {
    var nodes = getNodes(val);
    markNodes(nodes);
    var classNames = getClassNamesFromNodes(nodes);
    var styles = getStylesFromClassNames(classNames);
    var printedVal = printer(val);
    return (0, _replaceClassNames.replaceClassNames)(classNames, styles, printedVal, emotion.caches.key, classNameReplacer);
  }

  function test(val) {
    return val && !val.withEmotionStyles && (DOMElements ? isReactElement(val) || isDOMElement(val) : isReactElement(val));
  }

  function markNodes(nodes) {
    nodes.forEach(function (node) {
      node.withEmotionStyles = true;
    });
  }

  function getStylesFromClassNames(classNames) {
    var styles = '';
    // This could be done in a more efficient way
    // but it would be a breaking change to do so
    // because it would change the ordering of styles
    Object.keys(emotion.caches.registered).forEach(function (className) {
      var indexOfClassName = classNames.indexOf(className);
      if (indexOfClassName !== -1) {
        var nameWithoutKey = classNames[indexOfClassName].substring(emotion.caches.key.length + 1);
        // $FlowFixMe
        styles += emotion.caches.inserted[nameWithoutKey];
      }
    });
    var prettyStyles = void 0;
    try {
      prettyStyles = css.stringify(css.parse(styles));
    } catch (e) {
      console.error(e);
      throw new Error('There was an error parsing css in jest-emotion: "' + styles + '"');
    }
    return prettyStyles;
  }

  return { test: test, print: print };
}