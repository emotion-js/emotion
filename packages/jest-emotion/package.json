{
  "name": "jest-emotion",
  "version": "10.0.32",
  "description": "Jest utilities for emotion",
  "main": "dist/jest-emotion.cjs.js",
  "types": "types/index.d.ts",
  "files": [
    "src",
    "dist",
    "types",
    "serializer.js"
  ],
  "scripts": {
    "test:typescript": "dtslint types"
  },
  "dependencies": {
    "@babel/runtime": "^7.5.5",
    "@types/jest": "^23.0.2",
    "chalk": "^2.4.1",
    "css": "^2.2.1"
  },
  "devDependencies": {
    "@emotion/core": "^10.0.27",
    "dtslint": "^0.3.0",
    "emotion": "^10.0.27",
    "enzyme-to-json": "^3.2.1",
    "preact": "^8.2.9",
    "preact-render-to-json": "^3.6.6",
    "pretty-format": "^22.4.3",
    "react": "^16.5.2",
    "react-dom": "^16.5.2"
  },
  "author": "Kye Hohenberger",
  "homepage": "https://emotion.sh",
  "license": "MIT",
  "repository": "https://github.com/emotion-js/emotion/tree/master/packages/jest-emotion",
  "keywords": [
    "styles",
    "emotion",
    "react",
    "css",
    "css-in-js",
    "jest",
    "snapshot"
  ],
  "bugs": {
    "url": "https://github.com/emotion-js/emotion/issues"
  },
  "browser": {
    "./dist/jest-emotion.cjs.js": "./dist/jest-emotion.browser.cjs.js"
  }
}
