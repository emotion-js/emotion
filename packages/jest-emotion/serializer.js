const { createSerializer } = require('jest-emotion')
const emotion = require('emotion')

module.exports = createSerializer(emotion)
