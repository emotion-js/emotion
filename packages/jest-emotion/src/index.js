// @flow
import * as css from 'css'
import { replaceClassNames } from './replace-class-names'
import {
  getClassNamesFromNodes,
  isReactElement,
  isEmotionCssPropElementType,
  isEmotionCssPropEnzymeElement,
  isDOMElement,
  getStylesFromClassNames,
  getStyleElements,
  getKeys
} from './utils'

export { matchers } from './matchers'

function getNodes(node, nodes = []) {
  if (Array.isArray(node)) {
    for (let child of node) {
      getNodes(child, nodes)
    }
    return nodes
  }

  if (node.children) {
    for (let child of node.children) {
      getNodes(child, nodes)
    }
  }

  if (typeof node === 'object') {
    nodes.push(node)
  }

  return nodes
}

function getPrettyStylesFromClassNames(
  classNames: Array<string>,
  elements: Array<HTMLStyleElement>
) {
  let styles = getStylesFromClassNames(classNames, elements)

  let prettyStyles
  try {
    prettyStyles = css.stringify(css.parse(styles))
  } catch (e) {
    console.error(e)
    throw new Error(`There was an error parsing the following css: "${styles}"`)
  }
  return prettyStyles
}

type Options = {
  classNameReplacer?: (className: string, index: number) => string,
  DOMElements?: boolean
}

function filterEmotionProps(props = {}) {
  const {
    css,
    __EMOTION_TYPE_PLEASE_DO_NOT_USE__,
    __EMOTION_LABEL_PLEASE_DO_NOT_USE__,
    ...rest
  } = props

  rest.css = 'unknown styles'

  return rest
}

export function createSerializer({
  classNameReplacer,
  DOMElements = true
}: Options = {}) {
  let cache = new WeakSet()
  function print(val: *, printer: Function) {
    if (isEmotionCssPropEnzymeElement(val)) {
      return val.children.map(printer).join('\n')
    }
    if (isEmotionCssPropElementType(val)) {
      return printer({
        ...val,
        props: filterEmotionProps(val.props),
        type: val.props.__EMOTION_TYPE_PLEASE_DO_NOT_USE__
      })
    }
    const nodes = getNodes(val)
    const classNames = getClassNamesFromNodes(nodes)
    let elements = getStyleElements()
    const styles = getPrettyStylesFromClassNames(classNames, elements)
    nodes.forEach(cache.add, cache)
    const printedVal = printer(val)
    nodes.forEach(cache.delete, cache)
    let keys = getKeys(elements)
    return replaceClassNames(
      classNames,
      styles,
      printedVal,
      keys,
      classNameReplacer
    )
  }

  function test(val: *) {
    return (
      val &&
      ((!cache.has(val) &&
        (isReactElement(val) || (DOMElements && isDOMElement(val)))) ||
        isEmotionCssPropEnzymeElement(val) ||
        isEmotionCssPropElementType(val))
    )
  }
  return { test, print }
}

export let { print, test } = createSerializer()

export default { print, test }
