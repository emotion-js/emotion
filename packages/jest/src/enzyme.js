// @flow
export { createEnzymeSerializer } from './create-enzyme-serializer'
export { matchers } from './matchers'
