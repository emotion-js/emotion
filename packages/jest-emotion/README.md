# jest-emotion

> Jest testing utilities for emotion

# Installation

```bash
npm install --save-dev jest-emotion
```

# Snapshot Serializer

The easiest way to test React components with emotion is with the snapshot serializer. You can register the serializer via the `snapshotSerializers` configuration property in your jest configuration like so:

```js
// jest.config.js
module.exports = {
  // ... other config
  snapshotSerializers: ['jest-emotion']
}
```

Or you can customize the serializer via the `createSerializer` method like so: (the example below is with react-test-renderer but jest-emotion also works with enzyme and react-testing-library)

```jsx
import React from 'react'
import renderer from 'react-test-renderer'
import serializer from 'jest-emotion'
import styled from '@emotion/styled'

expect.addSnapshotSerializer(serializer)

test('renders with correct styles', () => {
  const H1 = styled.h1`
    float: left;
  `

  const tree = renderer.create(<H1>hello world</H1>).toJSON()

  expect(tree).toMatchSnapshot()
})
```

Refer to the [testing doc](https://github.com/emotion-js/emotion/blob/master/docs/testing.md) for more information about snapshot testing with emotion.

## Options

### `classNameReplacer`

jest-emotion's snapshot serializer replaces the hashes in class names with an index so that things like whitespace changes won't break snapshots. It optionally accepts a custom class name replacer, it defaults to the below.

```jsx
function classNameReplacer(className, index) {
  return `emotion-${index}`
}
```

```jsx
import { createSerializer } from 'jest-emotion'

expect.addSnapshotSerializer(
  createSerializer({
    classNameReplacer(className, index) {
      return `my-new-class-name-${index}`
    }
  })
)
```

### `DOMElements`

jest-emotion's snapshot serializer inserts styles and replaces class names in both React and DOM elements. If you would like to disable this behavior for DOM elements, you can do so by passing `{ DOMElements: false }`. For example:

```jsx
import { createSerializer } from 'jest-emotion'

// configures jest-emotion to ignore DOM elements
expect.addSnapshotSerializer(createSerializer({ DOMElements: false }))
```

# Custom matchers

## toHaveStyleRule

To make more explicit assertions when testing your styled components you can use the `toHaveStyleRule` matcher.

```jsx
import React from 'react'
import renderer from 'react-test-renderer'
import { matchers } from 'jest-emotion'
import styled from '@emotion/styled'

// Add the custom matchers provided by 'jest-emotion'
expect.extend(matchers)

test('renders with correct styles', () => {
  const Svg = styled('svg')`
    width: 100%;
  `

  const Div = styled('div')`
    float: left;
    height: 80%;
    &:hover {
      width: 50px;
    }
    ${Svg} {
      fill: green;
    }
    span {
      color: yellow;
    }
    @media screen and (max-width: 1200px) {
      font-size: 14px;
    }
  `

  const tree = renderer
    .create(
      <Div>
        <Svg />
        <span>Test</span>
      </Div>
    )
    .toJSON()

  expect(tree).toHaveStyleRule('float', 'left')
  expect(tree).not.toHaveStyleRule('height', '100%')
})
```

You can provide additional options for `toHaveStyleRule` matcher.  
`target` - helps to specify css selector or other component
where style rule should be found.

```js
expect(tree).toHaveStyleRule('width', '50px', { target: ':hover' })
```

```js
expect(tree).toHaveStyleRule('color', 'yellow', { target: 'span' })
```

```js
expect(tree).toHaveStyleRule('fill', 'green', { target: `${Svg}` })
```

`media` - specifies the media rule where the matcher
should look for the style property.

```js
expect(tree).toHaveStyleRule('font-size', '14px', {
  media: 'screen and (max-width: 1200px)'
})
```

Use `media` and `target` options to assert on rules within media queries and to target nested components, pseudo-classes, and pseudo-elements.

```jsx
import React from 'react'
import renderer from 'react-test-renderer'
import { matchers } from 'jest-emotion'
import styled from '@emotion/styled'

// Add the custom matchers provided by 'jest-emotion'
expect.extend(matchers)

test('renders with correct link styles', () => {
  const Container = styled.div`
    font-size: 14px;

    a {
      color: yellow;
    }

    a:hover {
      color: black;
    }

    @media (min-width: 768px) {
      font-size: 16px;
    }
  `

  const tree = renderer.create(<Container>hello world</Container>).toJSON()

  expect(tree).toHaveStyleRule('color', 'yellow', { target: /a$/ })
  expect(tree).toHaveStyleRule('color', 'black', { target: 'a:hover' })
  expect(tree).toHaveStyleRule('font-size', '16px', {
    media: '(min-width: 768px)'
  })
})
```

## Thanks

Thanks to [Kent C. Dodds](https://twitter.com/kentcdodds) who wrote [jest-glamor-react](https://github.com/kentcdodds/jest-glamor-react) which this library is largely based on. ❤️
