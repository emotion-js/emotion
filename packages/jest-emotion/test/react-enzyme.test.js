import 'test-utils/legacy-env'
/** @jsx jsx */
import * as enzyme from 'enzyme'
import { jsx } from '@emotion/core'
import { createSerializer as createEnzymeSerializer } from 'enzyme-to-json'
import { createSerializer } from 'jest-emotion'

expect.addSnapshotSerializer(createEnzymeSerializer())
expect.addSnapshotSerializer(createSerializer())

test('enzyme mount test', () => {
  const Greeting = ({ children }) => (
    <div css={{ backgroundColor: 'red' }}>{children}</div>
  )
  const tree = enzyme.mount(<Greeting>hello</Greeting>)
  expect(tree).toMatchSnapshot()
})

test('enzyme test with prop containing css element', () => {
  const Greeting = ({ children, content }) => (
    <div>
      {content} {children}
    </div>
  )

  const tree = enzyme.mount(
    <Greeting content={<p css={{ backgroundColor: 'blue' }}>Hello</p>}>
      World!
    </Greeting>
  )
  expect(tree).toMatchSnapshot()
})

test('enzyme test with prop containing css element not at the top level', () => {
  const Greeting = ({ children, content }) => (
    <div>
      {content} {children}
    </div>
  )

  const tree = enzyme.mount(
    <div>
      <Greeting
        content={
          <p id="something" css={{ backgroundColor: 'blue' }}>
            Hello
          </p>
        }
      >
        World!
      </Greeting>
    </div>
  )
  expect(tree).toMatchSnapshot()
})

test('enzyme test with prop containing css element with other props', () => {
  const Greeting = ({ children, content }) => (
    <div>
      {content} {children}
    </div>
  )

  const tree = enzyme.mount(
    <Greeting
      content={
        <p id="something" css={{ backgroundColor: 'blue' }}>
          Hello
        </p>
      }
    >
      World!
    </Greeting>
  )
  expect(tree).toMatchSnapshot()
})

test('enzyme test with prop containing css element with other label', () => {
  const Thing = ({ content, children }) => {
    return children
  }
  const Greeting = ({ children, content }) => (
    <Thing content={<div css={{ color: 'hotpink' }} />}>
      {content} {children}
    </Thing>
  )

  const tree = enzyme.mount(
    <Greeting
      content={
        <p id="something" css={{ backgroundColor: 'blue' }}>
          Hello
        </p>
      }
    >
      World!
    </Greeting>
  )
  expect(tree).toMatchSnapshot()
})
