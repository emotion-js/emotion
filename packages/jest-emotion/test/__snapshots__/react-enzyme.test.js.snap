// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`enzyme mount test 1`] = `
.emotion-0 {
  background-color: red;
}

.emotion-0 {
  background-color: red;
}

<Greeting>
  <div
    className="emotion-0"
  >
    hello
  </div>
</Greeting>
`;

exports[`enzyme test with prop containing css element 1`] = `
.emotion-0 {
  background-color: blue;
}

.emotion-0 {
  background-color: blue;
}

<Greeting
  content={
    <p
      css="unknown styles"
    >
      Hello
    </p>
  }
>
  <div>
    <p
      className="emotion-0"
    >
      Hello
    </p>
     
    World!
  </div>
</Greeting>
`;

exports[`enzyme test with prop containing css element not at the top level 1`] = `
.emotion-0 {
  background-color: blue;
}

.emotion-0 {
  background-color: blue;
}

<div>
  <Greeting
    content={
      <p
        css="unknown styles"
        id="something"
      >
        Hello
      </p>
    }
  >
    <div>
      <p
        className="emotion-0"
        id="something"
      >
        Hello
      </p>
       
      World!
    </div>
  </Greeting>
</div>
`;

exports[`enzyme test with prop containing css element with other label 1`] = `
.emotion-0 {
  background-color: blue;
}

.emotion-0 {
  background-color: blue;
}

<Greeting
  content={
    <p
      css="unknown styles"
      id="something"
    >
      Hello
    </p>
  }
>
  <Thing
    content={
      <div
        css="unknown styles"
      />
    }
  >
    <p
      className="emotion-0"
      id="something"
    >
      Hello
    </p>
     
    World!
  </Thing>
</Greeting>
`;

exports[`enzyme test with prop containing css element with other props 1`] = `
.emotion-0 {
  background-color: blue;
}

.emotion-0 {
  background-color: blue;
}

<Greeting
  content={
    <p
      css="unknown styles"
      id="something"
    >
      Hello
    </p>
  }
>
  <div>
    <p
      className="emotion-0"
      id="something"
    >
      Hello
    </p>
     
    World!
  </div>
</Greeting>
`;
